app/feature/feature-travel-certificate/src/main/kotlin/com/hedvig/android/feature/travelcertificate/ui/generate_when/TravelCertificateDateInputViewModel.kt

```diff
@@ -0,0 +1,231 @@
+  )
+
+internal class TravelCertificateDateInputPresenter(
+  private val contractId: String?,
+  private val getTravelCertificateSpecificationsUseCase: GetTravelCertificateSpecificationsUseCase,
+  private val createTravelCertificateUseCase: CreateTravelCertificateUseCase,
+) : MoleculePresenter<TravelCertificateDateInputEvent, TravelCertificateDateInputUiState> {
+  @Composable
+  override fun MoleculePresenterScope<TravelCertificateDateInputEvent>.present(
+    lastState: TravelCertificateDateInputUiState,
+  ): TravelCertificateDateInputUiState {
+    var loadIteration by remember { mutableIntStateOf(0) }
+
+    var generateIteration by remember { mutableIntStateOf(0) }
+
+    var screenContent by remember { mutableStateOf<DateInputScreenContent>(DateInputScreenContent.Loading) }
+
+    var travelDate by remember {
+      mutableStateOf(
+        Clock.System.now().toLocalDateTime(TimeZone.currentSystemDefault()).date,
+      )
+    }
+
+    var email by remember { mutableStateOf<ValidatedInput<String?>>(ValidatedInput(null)) }
+
+    CollectEvents { event ->
+      when (event) {
+        TravelCertificateDateInputEvent.RetryLoadData -> {
+          loadIteration++
+        }
+
+        is TravelCertificateDateInputEvent.ChangeEmailInput -> {
+          email = ValidatedInput(event.email)
+        }
+
+        is TravelCertificateDateInputEvent.ChangeDataInput -> {
+          travelDate = event.localDate
+        }
+
+        TravelCertificateDateInputEvent.GenerateTravelCertificate -> {
+          generateIteration++
+        }
+      }
+    }
+
+    LaunchedEffect(generateIteration) {
+      val currentContent = screenContent
+      val currentEmail = email.input
+      if (currentContent is DateInputScreenContent.Success && currentEmail != null) {
+        screenContent = DateInputScreenContent.Loading
+        createTravelCertificateUseCase.invoke(
+          contractId = currentContent.details.contractId,
+          startDate = travelDate,
+          isMemberIncluded = true,
+          coInsured = listOf(),
+          email = currentEmail,
+        ).fold(
+          ifLeft = { _ ->
+            screenContent = DateInputScreenContent.Failure
+            // todo: so on Retry we would just show the initial DateInput screen again
+          },
+          ifRight = { url ->
+            screenContent = DateInputScreenContent.UrlFetched(url)
+          },
+        )
+      }
+    }
+
+    LaunchedEffect(loadIteration) {
+      screenContent = DateInputScreenContent.Loading
+      getTravelCertificateSpecificationsUseCase
+        .invoke(contractId)
+        .fold(
+          ifLeft = { _ ->
+            // todo: so here we have 2 types of TravelCertificateError (Error and NotEligible),
+            // todo: but the second one shouldn't really appear in any case.
+            // todo: could probably change Either.Left type from TravelCertificateError to the usual ErrorMessage?
+            screenContent = DateInputScreenContent.Failure
```

Comment:
You could maybe use the same approach here with setting an errorMessage to the uiState.